Exercise 3

JoeOsborne22 opened the pull request:
Changes applied for module3 Testing

JoeOsborne22 committed:
exercise-3:
	- Integrated pytest into project
	- Added unit tests to cover and check some base functionality
	- enhanced application using a view model
	- update index.htmal to make use of new view model + section off
	  table creation into a macro for easy reusabillity

JoeOsborne22 committed:
exercise-3:
	- update id name on index.html

JoeOsborne22 committed:
exercise-3: Update showing all completed tasks to be only when there are
5 or less done tasks

JoeOsborne22 committed:
exercise-3: Initial attempt at integration tests

JoeOsborne22 committed:
exercise-3
	- Enhance app and test code to correctly work, load env vars and
	  mock request code
	- Add multiple asserts to integration test to ensure it works as
	  intended

JoeOsborne22 committed:
exercise-3:
	- Add initial e2e testing

JackMeadDev commented on README.md:
```diff
@@ -55,3 +55,26 @@ Now visit [`http://localhost:5000/`](http://localhost:5000/) in your web browser
+This to do app makes use of pytest for running test cases. Running the tests can be done in a few different ways shown below
+
+while being in the project directory (the test foleder should be visible or you should be within this dir) you can run the below command.
+This will run pytest on all available modaules/scrits
```
Small typo
```suggestion
This will run pytest on all available modules/scripts
```

JackMeadDev commented on geckodriver.log:
```diff
@@ -0,0 +1,107 @@
+1631032971180	geckodriver	INFO	Listening on 127.0.0.1:52508
```
Note that there's  no advantage to having this checked in - different users running the tests will end up adding merge conflicts for no real value, so I suggest removing it (`git rm --cached geckodriver.log`) and adding it to the .gitignore so it doesn't return

Similarly it's best to not track the geckodriver binary - it will need to vary between operating systems, it will potentially vary for different users (e.g. if on different versions of Firefox) and it's not tiny so sharing it between multiple projects is easier - we can rely on our tests finding the driver on the PATH rather than directly in the folder

JackMeadDev commented on tests/test_ToDoApp.py:
```diff
@@ -0,0 +1,107 @@
+
+	view = taskCards(items)
+	
+	todo_items_tasks=[]
+	for card in view.todo_items:
+    		todo_items_tasks.append(card.title)
+	assert todo_items_tasks == ['task 1']
```
Minor, but we could capture this slightly more succinctly with e.g.:
```suggestion
	assert len(view.todo_items) == 1
	assert view.todo_items[0].title == 'task 1'
```

JackMeadDev commented on tests/test_integrationToDoApp.py:
```diff
@@ -0,0 +1,111 @@
+        mock_resp = Mock()
+        # mock raise_for_status call w/optional error
+        mock_resp.raise_for_status = Mock()
+        if raise_for_status:
+            mock_resp.raise_for_status.side_effect = raise_for_status
```
This will never be triggered since `raise_for_status` is always None - this is currently unused so I'd  be inclined to get rid of it

JackMeadDev commented on todo_app/viewModel.py:
```diff
@@ -0,0 +1,57 @@
+        return list
+        
+    @property
+    def should_show_all_done_items(self):
```
Nice work on the stretch goals!

JackMeadDev commented on todo_app/templates/index.html:
```diff
@@ -25,32 +56,29 @@ <h2>Add New Item</h2>
+    {{ display_table('Doing', viewModel.doing_items, 'doing') }}
+  {% endif %}
+
+<!-- Only display all the done items if we have 3 or less -->
```
Note that this comment is wrong in exactly the sort of way comments often do get out of date! The `should_show_all_done_items` name gives developers a clear place to investigate if they need to care why it's showing/not showing, but hides the details to let them change safely if needed

JackMeadDev commented on tests_e2e/test_e2e.py:
```diff
@@ -0,0 +1,64 @@
+        yield driver
+
+
+def test_mainPage(driver):
```
I think you're currently running this test against a locally hosted version of the app, rather than having the test spin up the app for you - does that sound right?
The reason I think that is you're not currently using the app_with_temp_board function that you defined above and that I can see is working - you need to specify it as a fixture you're using by mentioning it in the test's arguments:
```python
def test_mainPage(driver, app_with_temp_board):
    driver.get('http://localhost:5000/')
    assert driver.title == 'To-Do App'        
```

Without that your test fails for me, with it it passes (with the other twist above)

JackMeadDev commented on tests_e2e/test_e2e.py:
```diff
@@ -0,0 +1,64 @@
+@pytest.fixture(scope='module')
+def app_with_temp_board():
+    # Create the new board & update the board id environment variable
+    board_id = (createTrelloBoard('Selenium_Test_Board'))["id"]
```
Note that this requires your environment variables in order to talk to Trello, which aren't set up with Pytest by default. Instead you will need to import the dotenv functions as with your integration test and add a couple of lines before this to find & load your `.env` file (rather than `.env.test`) so that the test has the right credentials to talk to Trello, i.e.:
```python
    file_path = find_dotenv('.env')
    load_dotenv(file_path, override=True)
```

JackMeadDev commented on tests_e2e/test_e2e.py:
```diff
@@ -0,0 +1,64 @@
+
+
+def test_mainPage(driver):
+    driver.get('http://localhost:5000/')
+    assert driver.title == 'To-Do App'          
```
This checks the page loads, but we should check some basic functionality, e.g. creating a to-do and checking it appears. Unchecked code below, but that might look something like:
```python
    name_input = driver.find_element_by_id('<id for input>')
    name_input.send_keys('Test item')
    submit_button = driver.find_element_by_id('<id for button>')
    submit_button.click()
    new_item_status = driver.find_element_by_class_name('<class for items in todo status>')
    assert new_item_status is not None
```

JackMeadDev commented on README.md:
```diff
@@ -55,3 +55,26 @@ Now visit [`http://localhost:5000/`](http://localhost:5000/) in your web browser
 Then generate an API key and token by following the instructions here - https://trello.com/app-key
+
+
+## Testing The app
```
Minor, but we should also mention that this has a dependency on having Firefox installed (and Geckodriver if you do remove that from the checked in files)

JackMeadDev reviewed (changes requested):
Nice work, your app was easy to run and your tests all look good - the unit/integration tests are great examples of their kind, but the end-to-end test is not quite working out of the box so we should fix that up - more detailed comments are below along with some optional suggestions, but the important bit for that is that:
* It runs without having the app running in the background (i.e. the test should spin up the app)
* It has some sort of page interaction - see the suggestion below for more info

JoeOsborne22 committed:
exercise-3
	Implemented reccomended changes from PR
	Main change is around fixing and enhancing e2e tests

JackMeadDev commented on tests_e2e/test_e2e.py:
```diff
@@ -26,17 +27,27 @@ def createTrelloBoard(boardName):
-    os.environ['trelloBoardID'] = board_id
+    board = (createTrelloBoard('Selenium_Test_Board'))
+    board_id = board["id"]
+    file_path = find_dotenv('.env')
+    load_dotenv(file_path, override=True)
```
Note that this currently happens _after_ you've tried to use these variables (in creating the board on the previous two lines)! If we shift this stage up to the top of this function then this test should be good to go 😄

JackMeadDev reviewed (commented):
Nice work, I'm happy this is merged but note that it doesn't _quite_ just work still! I've left one comment below to point out why - it's just a case of reordering two lines; I'm happy for you to make that change as part of a future module if easiest.
